crates/neon/src/thread/mod.rs

```diff
@@ -0,0 +1,232 @@
+//! instances:
+//!
+//! ```
+//! # use neon::prelude::*;
+//! # use neon::thread::LocalKey;
+//! static THREAD_ID: LocalKey<u32> = LocalKey::new();
+//!
+//! pub fn thread_id<'cx, C: Context<'cx>>(cx: &mut C) -> NeonResult<u32> {
+//!     THREAD_ID.get_or_try_init(cx, |cx| {
+//!         let global = cx.global();
+//!         let require: Handle<JsFunction> = global.get(cx, "require")?;
+//!         let worker: Handle<JsObject> = require.call_with(cx)
+//!             .arg(cx.string("node:worker_threads"))
+//!             .apply(cx)?;
+//!         let threadId: Handle<JsNumber> = worker.get(cx, "threadId")?;
+//!         Ok(threadId.value(cx) as u32)
+//!     }).cloned()
+//! }
+//! ```
+//!
+//! ### The Addon Lifecycle
+//!
+//! For some use cases, a single shared global constant stored in a `static` variable
+//! might be sufficient:
+//!
+//! ```
+//! static MY_CONSTANT: &'static str = "hello Neon";
+//! ```
+//!
+//! This variable will be allocated when the addon is first loaded into the Node.js
+//! process. This works fine for single-threaded applications, or global immutable
+//! data.
+//!
+//! However, since the addition of [worker threads][workers] in Node v10,
+//! modules can be instantiated multiple times in a single Node process. This means
+//! that while the dynamically-loaded binary library (i.e., the Rust implementation of
+//! the addon) is only loaded once in the running process, but its `main()` function
+//! is executed multiple times with distinct module objects, once per application thread:
+//!
+//! ![The Node.js addon lifecycle, described in detail below.][lifecycle]
+//!
+//! This means that any thread-local data needs to be initialized separately for each
+//! instance of the addon. This module provides a simple container type, [`LocalKey`](LocalKey),
+//! for allocating and initializing thread-local data. (Technically, this data is stored in the
+//! addon's [module instance][environment], which is equivalent to being thread-local.)
+//!
+//! A common example is when an addon defines a custom datatype, which cannot be used across
+//! separate threads. The constructor for the datatype can be [rooted](crate::handle::Root) and
+//! saved in thread-local storage:
+//!
+//! ```
+//! # use neon::prelude::*;
+//! # use neon::thread::LocalKey;
+//! # fn initialize_my_datatype<'cx, C: Context<'cx>>(cx: &mut C) -> JsResult<'cx, JsFunction> { unimplemented!() }
+//! static MY_CONSTRUCTOR: LocalKey<Root<JsFunction>> = LocalKey::new();
+//!
+//! pub fn my_constructor<'cx, C: Context<'cx>>(cx: &mut C) -> JsResult<'cx, JsFunction> {
+//!     let constructor = MY_CONSTRUCTOR.get_or_try_init(cx, |cx| {
+//!         let constructor: Handle<JsFunction> = initialize_my_datatype(cx)?;
+//!         Ok(constructor.root(cx))
+//!     })?;
+//!     Ok(constructor.to_inner(cx))
+//! }
+//! ```
+//!
+//! Notice that if this code were implemented without a `LocalKey`, it would fail whenever
```

Comment:
```suggestion
//! Notice that if this code were implemented without a `LocalKey`, it would panic whenever
```

As an aside, it looks like I didn't update the `Root` documentation when this behavior was added.

Reply:
I went ahead and added panic notes to `Root::{into_inner, to_inner}`